gamemodes/cinema/gamemode/modules/theater/services/sh_bitchute.lua

```diff
@@ -0,0 +1,84 @@
+local SERVICE = {}
+
+SERVICE.Name 		= "Bitchute"
+
+SERVICE.NeedsCodecs = true
+
+function SERVICE:GetKey( url )
+	match = string.match(url.path,"/.+/(.+[^/])")
+	if match != nil and string.find(url.host,"bitchute.com") then
+		return match
+	end
+	return false
+end
+
+if CLIENT then
+	function SERVICE:GetVideoInfoClientside(key, callback)
+		EmbeddedCheckCodecs(function()
+			vpanel = vgui.Create("DHTML")
+		
+			vpanel:SetSize(100,100)
+		
+			vpanel:SetAlpha(0)
+		
+			vpanel:SetMouseInputEnabled(false)
+			
+			timer.Simple(20,function() 
+				if IsValid(vpanel) then
+					vpanel:Remove() 
+					print("Failed")
+					callback()
+				end
+			end)
+		
+			function vpanel:ConsoleMessage(msg)
+				if msg:StartWith("DURATION:") then
+					local duration = math.ceil(tonumber(string.sub(msg,10)))
+					print("Duration: "..duration)
+					self:Remove()
+					print("Success!")
+					callback({duration=duration})
+				end
+			end
+			
+			http.Fetch("https://www.bitchute.com/embed/"..key,
+				function(body,length,headers,code)
+					match = string.match(body,"source src=\"(.+)\" type")
+					if match != nil then
+						vpanel:OpenURL("http://swampservers.net/cinema/filedata.php?file="..match)
+					end
+				end,
+				function(err)
+					print("Failed to reach bitchute page while fetching duration")
+					vpanel:Remove()
+					callback()
+				end
+			)
+		end,
+		function()
+			chat.AddText("You need codecs to request this. Press F2.")
+			return callback()
+		end)
+	end
+	
+	function SERVICE:LoadVideo( Video, panel )
+		http.Fetch("https://www.bitchute.com/embed/"..Video:Key(),
```

Comment:
LoadVideo is run clientside for everyone watching. Can this function be pushed back to serverside and run once, or is the resulting link unique to each client?

Reply:
it's not unique to each client, I'm not sure how you'll implement it serverside

Reply:
I added the serverside of youtube and soundcloud. Youtube is an example of how to use the server to get extra data asychronously; you can use the .data field of the video to store anything arbitrary other than title/duration/thumbnail (which is why I renamed the identifier the "KEY", so I could use "DATA" as the name for the other stuff). The soundcloud serverside is an example of the plumbing needed to use the client for stuff

Reply:
I still need access to the info table in EmbeddedCheckCodecs unless you're fine with using a http function in there

Reply:
Not sure what you mean. If you update the PR to what you think would work, I can let you know if there's a proper way to do it

Reply:
you talked about putting the real url into the .data member/whatever but I can't access that where I actually need it

Reply:
Oh, sorry, so if you look at sv_youtube.lua there's the whole sv_GetVideoInfo.youtube which is the serverside way of getting video data before it's going to play. In that function it populates the "info" table. You can use info.data there to store arbitrary stuff.

Reply:
yes but what good is that when I can't access it anywhere else

Reply:
unless info.key is the same variable being passed to both GetVideoInfoClientside and LoadVideo

Reply:
The Video object should have a Data() field to get it when you call service:LoadVideo. I think a good way to handle this would be to store the actual URL which gets eventually passed to the JWPlayer (ie "file.html") in the Data field instead of using the Key field for it. For this service it's optional, but for the other service you still need to make your asynchronous request inside of a sv_GetVideoInfo function (because GetKey is not asychronous and it won't work the way it is right now)

Reply:
what about the GetVideoInfoClientside function, only the key and callback is passed to it

Reply:
Actually, you're right, if you're doing a GetVideoInfoClientside then with the current plumbing, you can only send the Key to the client. If you were using the Data as the video URL, well, it wouldn't work unless I added support for that.

You can just make HTTP requests in GetVideoInfoClientside like it currently is, its fine. The blocker on this right now is the other comment about the asynchronous return not working